File: scripts/convert-dependence-by-level.py
"""
Convert data/Commo_Dep_by_Country_Group(Right).xlsx
  → public/assets/data/cdde_dependence_by_level.csv

Excel structure (row 1 = header, rows 2+ = data):
  Col A: Name   — full country group name
  Col B: Value  — mean commodity export share (0–100)

The short group label and economy count are derived from a fixed mapping
since the Excel contains only the average value.
"""

import csv
import sys
from pathlib import Path

try:
    import openpyxl
except ImportError:
    import subprocess
    print("openpyxl not found — installing...")
    subprocess.check_call([sys.executable, "-m", "pip", "install", "openpyxl",
                           "--break-system-packages", "-q"])
    import openpyxl

# Maps full Excel name → (short label used in chart, economy count)
# Bar colour is derived in the component: "Developed" → green, all others → blue.
GROUP_MAP = {
    "Small island developing states":   ("SIDS",             37),
    "Landlocked developing countries":  ("LLDCs",            32),
    "Least developed countries":        ("LDCs",             44),
    "Other developing countries":       ("Other developing", 55),
    "Developed countries":              ("Developed",        48),
}


def convert(xlsx_path: Path, csv_path: Path) -> None:
    wb = openpyxl.load_workbook(xlsx_path, read_only=True, data_only=True)
    ws = wb.active

    rows_written = 0
    csv_path.parent.mkdir(parents=True, exist_ok=True)

    with csv_path.open("w", newline="", encoding="utf-8") as fh:
        writer = csv.writer(fh)
        writer.writerow(["group", "economies", "avg_pct"])

        first = True
        for row in ws.iter_rows(values_only=True):
            name, value, *_ = list(row) + [None]
            if first:
                first = False
                continue
            if name is None or value is None:
                continue
            name = str(name).strip()
            mapping = GROUP_MAP.get(name)
            if not mapping:
                print(f"  Warning: unknown group '{name}' — skipped")
                continue
            short, economies = mapping
            writer.writerow([short, economies, float(value)])
            rows_written += 1

    print(f"Written {rows_written} rows → {csv_path}")


if __name__ == "__main__":
    root = Path(__file__).parent.parent
    convert(
        root / "data" / "Commo_Dep_by_Country_Group(Right).xlsx",
        root / "public" / "assets" / "data" / "cdde_dependence_by_level.csv",
    )

File: scripts/convert-dependence-by-level.mjs
/**
 * Converts data/Avg by level(MAPS).xlsx
 * → public/assets/data/cdde_dependence_by_level.csv
 *
 * Expected Excel columns:
 *   A: group      — country group label  (use "Developed" for the developed group — bar turns green)
 *   B: economies  — number of economies in the group
 *   C: avg_pct    — mean commodity export share, 0–100  (e.g. 79.7)
 *
 * Example rows:
 *   LLDCs   32  79.7
 *   LDCs    44  79.5
 *   SIDS    38  65.9
 *   Other developing  89  61.3
 *   Developed         40  36.0
 *
 * Note: the bar colour is derived automatically — "Developed" → green, all others → blue.
 */

import { createRequire } from 'module';
import { createWriteStream } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

const require = createRequire(import.meta.url);
const XLSX = require('xlsx');

const ROOT   = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const INPUT  = resolve(ROOT, 'data', 'Avg by level(MAPS).xlsx');
const OUTPUT = resolve(ROOT, 'public', 'assets', 'data', 'cdde_dependence_by_level.csv');

const wb    = XLSX.readFile(INPUT);
const sheet = wb.Sheets[wb.SheetNames[0]];
const rows  = XLSX.utils.sheet_to_json(sheet, { header: 1, defval: '' });

const LABEL_RENAME = {
  'Other developing': 'Other developing economies',
  'Developed':        'Developed economies',
};

const header = ['group', 'economies', 'avg_pct'];
const out    = [header.join(',')];

for (let i = 1; i < rows.length; i++) {
  const [rawGroup, economies, avg_pct] = rows[i].map(c => String(c || '').trim());
  if (!rawGroup || !avg_pct) continue;
  const group = LABEL_RENAME[rawGroup] ?? rawGroup;
  out.push([group, +economies, +avg_pct].join(','));
}

createWriteStream(OUTPUT).end(out.join('\r\n') + '\r\n');
console.log(`✓ ${out.length - 1} rows → ${OUTPUT}`);
